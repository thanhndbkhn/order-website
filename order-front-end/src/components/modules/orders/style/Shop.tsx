import styled from 'styled-components'

export const ListItemProduct = styled.div`
  width: 60%;
  float: left;
  display: flex;
  flex-wrap: wrap;
  overflow-y: auto;
  @media (max-width: 1024px) {
    width: 100%;
    display: flex;
    flex-wrap: wrap;
  }
`
export const OrderList =styled.div`
  width: 40%;
  float: right;
  @media (max-width: 1024px) {
    float: none;
    width: 100%;
  }
`
export const CardProduct = styled.div`
  width: 25%;
  flex: auto;
  position: relative;
  min-height: 1px;
  padding-right: 15px;
  padding-left: 15px;
`
export const ProductWrapper = styled.div`
  border: 1px solid #d8d8b6;
  overflow: hidden;
  margin-bottom: 30px;
  position: relative;
  padding: 5px;
  text-align: center;
  h2 {
    color: #fe980f;
  }
  img {
    width: 100%;
  }
`

