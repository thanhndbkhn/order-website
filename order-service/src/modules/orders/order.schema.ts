import { Prop, Schema, SchemaFactory } from '@nestjs/mongoose';
import { IsArray, IsEnum, IsInt, IsNotEmpty, IsNumber, IsPositive, IsString } from 'class-validator';
import { Document } from 'mongoose';
import { ORDER_STATUS } from './constants/orders.constants';
import { OrderItem } from './interfaces/orderItem.interface';

export type OrderDocument = Order & Document;

@Schema({ timestamps: true , toJSON: {transform : (doc, ret, options) => {
  const order = {...ret, orderId : ret._id}
  ret.orderId = ret._id;
  return order;
}}})

export class Order {

  @Prop({ autoCreate: true, id: true, autoIndex: true, auto: true })
  orderId: string;

  @Prop({ required: true })
  @IsInt()
  @IsNotEmpty()
  userId: number;

  @Prop({ default: ORDER_STATUS.CREATED })
  @IsNotEmpty()
  @IsEnum(ORDER_STATUS)
  orderStatus: ORDER_STATUS;

  @Prop()
  @IsNumber()
  @IsNotEmpty()
  @IsPositive()
  totalPrice: number;
  
  @IsString()
  description: string;

  @Prop({ autoCreate: true })
  @IsArray()
  orderItems: OrderItem[];
}

export const OrderSchema = SchemaFactory.createForClass(Order);