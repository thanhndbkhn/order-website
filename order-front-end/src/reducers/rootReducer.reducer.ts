import { combineReducers } from 'redux';
import orders, {OrderState} from '../components/modules/orders/reducers/orders.reducers';

const rootReducer = combineReducers<IRootState>({
    orders
})

export default rootReducer;
export interface IRootState {
    orders: OrderState
}