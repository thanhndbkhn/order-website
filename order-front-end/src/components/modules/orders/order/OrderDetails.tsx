import { connect } from 'react-redux';
import React, { useState, useEffect } from 'react';
import { IRootState } from '../../../../reducers/rootReducer.reducer';
import { getOrderDetails, cancelOrder, OrderAction } from '../reducers/orders.reducers';
import * as R from 'ramda';
import { MESSAGE_NOTIFICATION, ORDER_STATUS } from '../constants/order.constants';
import { OrderInfo, ListProductItem, PopupBody, PopupDetail, PopupHeader } from '../style/Order';
import { ButtonCancel, ButtonConfirm } from '../style/Button';
import {ToastMessage} from '../../../common/ToastMessage';
import { statusToText } from '../helper/helper';
import { Action, TOAST_BACKGROUND_COLOR } from '../../../../constants/common.constants';
import { ItemOrder } from '../interface/itemOrder.interface';
export interface IOrderDetailsProps extends DispatchProps, StateProps {
  orderId: string;
  closeOrderDetails?: () => void;
}

const OrderDetails: React.FC<IOrderDetailsProps> = (props) => {

  const [status, setStatus] = useState(Action.None);

  useEffect(() => {
    props.getOrderDetails(props.orderId);
    setStatus(Action.None);
    // eslint-disable-next-line
  }, [])

  useEffect(() => {
    if (props.actionStatus.status === OrderAction.CancelSuccess) {
      props.getOrderDetails(props.orderId);
      setStatus(Action.Success)
    }
    if (props.actionStatus.status === OrderAction.CancelError) {
      props.getOrderDetails(props.orderId);
      setStatus(Action.Error)
    }
    // eslint-disable-next-line
  }, [props.actionStatus]);

  const cancelOrder = () => {
    props.cancelOrder(props.orderId)
  }

  return (
    <PopupDetail>
      <ToastMessage
        message={status === Action.Success ? MESSAGE_NOTIFICATION.CANCELED_SUCCESS : MESSAGE_NOTIFICATION.CANCELED_ERROR}
        isDisplay={status !== Action.None ?? false}
        color={status === Action.Success ? TOAST_BACKGROUND_COLOR.SUCCESS : TOAST_BACKGROUND_COLOR.ERROR} />
      <PopupBody >
        <PopupHeader data-testid="detail-order">
          Order Details
        </PopupHeader>
        {props.orderDetails &&
          <div>
            {props.orderItems && props.orderItems.map((item: ItemOrder, idx) => (
              <ListProductItem key={idx}>
                <img src={item.product.imageUrl} alt="" />
                <div>
                  <h3>{item.product.productName}</h3>
                </div>
                <div>
                  <h2>{item.quantity}</h2>
                </div>
                <div>
                  <span>$ {item.product.price}</span>
                  {/* <time className="recent-time">Sep 25, 2019</time> */}
                </div>
              </ListProductItem>
            ))}
            <OrderInfo>
              <div>
                <h3>Total : {R.path(['totalPrice'], props.orderDetails)}$</h3>
              </div>
              <div>
                <h2>Status: {statusToText(R.path(['orderStatus'], props.orderDetails) ?? 0)}</h2>
              </div>
              <div>
                <ButtonConfirm onClick={props.closeOrderDetails}>OK</ButtonConfirm>
                <span>  </span>
                {R.path(['orderStatus'], props.orderDetails) === ORDER_STATUS.CONFIRMED &&
                  <ButtonCancel onClick={cancelOrder}>Cancel</ButtonCancel>
                }
              </div>
            </OrderInfo>
          </div>}
      </PopupBody>
    </PopupDetail>
  )
}


const mapStateToProps = ({ orders }: IRootState) => ({
  actionStatus: orders.actionStatus,
  orderDetails: orders.orderDetails,
  orderItems: orders.orderItems
});

const mapDispatchToProps = {
  getOrderDetails,
  cancelOrder
}

type DispatchProps = typeof mapDispatchToProps;

type StateProps = ReturnType<typeof mapStateToProps>;

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(OrderDetails);

