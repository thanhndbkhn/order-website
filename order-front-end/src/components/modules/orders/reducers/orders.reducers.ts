import { Dispatch } from 'redux';
import axios from 'axios';
import { userIdLogin } from '../constants/order.constants';

export const ACTION_TYPES = {
  CREATE_ORDER: 'create-order',
  GET_LIST_ORDER: 'get-list-order',
  GET_DETAIL_ORDER: 'get-order-detail',
  CANCEL_ORDER: 'cancel-order'
}

export enum OrderAction {
  None = 'None',
  Success = 'Success',
  CreateSuccess = 'Create Success',
  CancelSuccess = 'CancelSuccess',
  GetDetailSuccess = 'GetDetailSuccess',
  GetListSuccess = 'GetListSuccess',
  CancelError = 'CancelError',
  CreateError = 'CreateError',
}

const initialState = {
  orderCreated: null,
  actionStatus: { status: OrderAction.None },
  orderList: [],
  orderDetails: null,
  orderItems: [],
}

export type OrderState = Readonly<typeof initialState>;

// Reducer
const ordersReducer = (state: OrderState = initialState, action: any): OrderState => {
  switch (action.type) {
    case ACTION_TYPES.CREATE_ORDER:
      return {
        ...state,
        actionStatus: action.response.data ? { status: OrderAction.CreateSuccess } : { status: OrderAction.CreateError }
      }
    case ACTION_TYPES.GET_LIST_ORDER:
      return {
        ...state,
        orderList: action.response.data
      }

    case ACTION_TYPES.GET_DETAIL_ORDER:
      return {
        ...state,
        orderDetails: action.response.data,
        orderItems: action.response.data.orderItems,
        actionStatus: { status: OrderAction.GetListSuccess }
      }

    case ACTION_TYPES.CANCEL_ORDER:
      return {
        ...state,
        actionStatus: action.response.data ? { status: OrderAction.CancelSuccess } : { status: OrderAction.CancelError }
      }

    default:
      return {
        ...state,
      }
  }
}

export default ordersReducer;

export const createOrder: (order: any) => void = order => async (dispatch: Dispatch) => {
  const response = await axios.post(
    process.env.REACT_APP_ORDER_API + ``,
    order,
    { headers: { 'Content-Type': 'application/json' } }
  ).catch(err => {
    return err;
  });

  dispatch({
    type: ACTION_TYPES.CREATE_ORDER,
    response: response
  });
}

export const getListOder: (limit: number, offset: number) => void = (limit, offset) => async (dispatch: Dispatch) => {
  const response = await axios.get(
    process.env.REACT_APP_ORDER_API + ``,
    {
      params: {
        userId: userIdLogin,
        limit,
        offset
      }
    }
  )

  dispatch({
    type: ACTION_TYPES.GET_LIST_ORDER,
    response: response
  });
}

export const getOrderDetails: (orderId: string) => void = (orderId) => async (dispatch: Dispatch) => {
  const response = await axios.get(
    process.env.REACT_APP_ORDER_API + `/${orderId}`,
  );

  dispatch({
    type: ACTION_TYPES.GET_DETAIL_ORDER,
    response: response
  });
}

export const cancelOrder: (orderId: string) => void = (orderId) => async (dispatch: Dispatch) => {
  const response = await axios.put(
    process.env.REACT_APP_ORDER_API + `/${orderId}/cancel`,
  ).catch(err => {
    return err;
  });

  dispatch({
    type: ACTION_TYPES.CANCEL_ORDER,
    response: response
  });
}