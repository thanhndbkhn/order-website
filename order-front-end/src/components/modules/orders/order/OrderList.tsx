import { connect } from 'react-redux';
import React, { useState, useEffect } from 'react';
import { IRootState } from '../../../../reducers/rootReducer.reducer';
import { getListOder, OrderAction } from '../reducers/orders.reducers';
import { LIMIT, OFFSET, TIME_UPDATE_DATA_ORDER } from '../constants/order.constants';
import { Order } from '../interface/order.interface';
import { ItemOrder } from '../interface/itemOrder.interface';
import Moment from 'moment';
import OrderDetails from './OrderDetails'
import { TableOrderList } from '../style/Table';
import { ListProductName } from '../style/Order';
import { OrderStatus } from './OrderStatus';
import { Pagination } from '../../../common/Pagination';
export interface IOrderListProps extends DispatchProps, StateProps {

}

const OrderList: React.FC<IOrderListProps> = (props) => {
  const [idOrderOpen, setIdOrderOpen] = useState('');
  const [offset, setOffset] = useState(OFFSET);

  useEffect(() => {
    props.getListOder(LIMIT, OFFSET);
    // eslint-disable-next-line
  }, [])

  useEffect(() => {
    const interval = setInterval(() => props.getListOder(LIMIT, offset), TIME_UPDATE_DATA_ORDER);
    return (() => {
      clearInterval(interval);
    });
    // eslint-disable-next-line
  }, [offset])


  useEffect(() => {
    if (props.actionStatus.status === OrderAction.CancelSuccess) {
      props.getListOder(LIMIT, offset);
    }
    // eslint-disable-next-line
  }, [props.actionStatus])

  const formatDate = (date: Moment.MomentInput) => {
    return Moment(date).format('DD/MM/yyyy')
  }

  const openOrderDetail = (order: any) => {
    setIdOrderOpen(order._id);
  }

  const changePageNumber = (offset: number) => {
    props.getListOder(LIMIT, offset);
    setOffset(offset);
  }

  return (<>
    <Pagination changePageNumberCb={changePageNumber} />
    {props.orderList && props.orderList.length > 0 &&
      <>
        <TableOrderList>
          <thead>
            <tr>
              <th>ListProduct</th>
              <th>Total payment</th>
              <th>Date Create</th>
              <th>Status</th>
            </tr>
          </thead>
          <tbody>
            {props.orderList.map((order: Order, idx) => (
              <tr key={idx}>
                <td>
                  <div>
                    <ListProductName data-testid={order.orderId} onClick={() => openOrderDetail(order)}>
                      {order?.orderItems?.map((itemOrder: ItemOrder, index) => (
                        // eslint-disable-next-line
                        <a key={index}>
                          {index > 0 ? ', ' : ''}
                          {itemOrder.product?.productName}
                        </a>
                      ))}</ListProductName>
                  </div>
                </td>
                <td>
                  <div>${order.totalPrice}</div>
                </td>
                <td>
                  <div>{formatDate(order.createdAt)}</div>
                </td>
                <td>
                  <OrderStatus orderStatus={order.orderStatus} />
                </td>
              </tr>
            ))}
          </tbody>
        </TableOrderList>
      </>
    }
    {idOrderOpen &&
      <OrderDetails
        orderId={idOrderOpen}
        closeOrderDetails={() => setIdOrderOpen('')} />}
  </>);
}

const mapStateToProps = ({ orders }: IRootState) => ({
  actionStatus: orders.actionStatus,
  orderList: orders.orderList
});

const mapDispatchToProps = {
  getListOder
}

type DispatchProps = typeof mapDispatchToProps;

type StateProps = ReturnType<typeof mapStateToProps>;

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(OrderList);
