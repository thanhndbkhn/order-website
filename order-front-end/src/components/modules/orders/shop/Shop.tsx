
import React, { useState, useEffect } from 'react';
import { DATA_PRODUCT_DUMMY } from '../data-dummy/product';
import { Product } from '../interface/product.interface';
import { connect } from 'react-redux';
import _ from 'lodash';
import { Order } from '../interface/order.interface';
import {OrderItem} from './OrderItem';
import { IRootState } from '../../../../reducers/rootReducer.reducer';
import { createOrder, OrderAction } from '../reducers/orders.reducers';
import { MESSAGE_NOTIFICATION, QUANTITY_NUMBER_ADD_DEFAULT, userIdLogin } from '../constants/order.constants';
import { ListItemProduct, OrderList } from '../style/Shop';
import { Button } from '../style/Button';
import { TableOrderItem } from '../style/Table';
import {ProductCard} from './ProductCard';
import {ToastMessage} from '../../../common/ToastMessage';
import { Action, TOAST_BACKGROUND_COLOR } from '../../../../constants/common.constants';

export interface IShopProps extends StateProps, DispatchProps {

}

const Shop: React.FC<IShopProps> = (props) => {

  const listProduct = DATA_PRODUCT_DUMMY;

  const [order, setOrder] = useState<Order>();
  const [status, setStatus] = useState(Action.None);

  useEffect(() => {
    if (props.actionStatus.status === OrderAction.CreateSuccess) {
      setOrder(undefined);
      setStatus(Action.Success);
    }
  }, [props.actionStatus]);

  useEffect(() => {
    setStatus(Action.None);
  }, [])

  const addProductToOrder = (product: Product) => {
    let orderCreate = null;
    if (order) {
      order.userId = userIdLogin;
      const productExists = order.orderItems.find(item => item.product.productId === product.productId);
      if (productExists) {
        order.orderItems.map(item => {
          if (item.product.productId === product.productId) {
            item.quantity += QUANTITY_NUMBER_ADD_DEFAULT;
          }
          return item;
        })
      } else {
        order.orderItems.push({ product, quantity: QUANTITY_NUMBER_ADD_DEFAULT })
      }

      orderCreate = { ...order };
    } else {
      orderCreate = {
        userId: userIdLogin,
        orderItems: [{ product, quantity: QUANTITY_NUMBER_ADD_DEFAULT }]
      }
    }
    setOrder(orderCreate);
  }

  const removeProductFromOrder = (product: Product) => {
    const idxRemove = order?.orderItems?.findIndex(elm => elm.product.productId === product.productId);

    if (order && !_.isNil(idxRemove) && idxRemove >= 0) {
      const orderUpdate = { ...order };
      orderUpdate.orderItems.splice(idxRemove, 1);
      setOrder(orderUpdate);
    }
  }

  const getTotalPrice = () => {
    let price = 0;
    order?.orderItems.forEach(item => {
      price += item.product.price * item.quantity;
    })
    return price;
  }

  const orderProduct = () => {
    setStatus(Action.None);
    let orderCreate = { ...order }
    orderCreate.totalPrice = getTotalPrice();
    props.createOrder(orderCreate);
  }

  return (<>
    <ToastMessage
      message={MESSAGE_NOTIFICATION.CREATED_SUCCESS}
      isDisplay={status !== Action.None ?? false}
      color={status === Action.Success ? TOAST_BACKGROUND_COLOR.SUCCESS : ''} />
    <ListItemProduct>
      {listProduct.map((product, index) => (
        <ProductCard
          key={index}
          product={product}
          addProductToOrder={addProductToOrder}
        />
      ))}
    </ListItemProduct>
    <OrderList>
      <TableOrderItem>
        <thead>
          <tr>
            <td>Item</td>
            <td>Description</td>
            <td>Price</td>
            <td>Total</td>
            <td></td>
          </tr>
        </thead>
        {order?.orderItems && order.orderItems.length > 0 &&
          <tbody>
            {order.orderItems.map((item, idx) => (
              <OrderItem
                key={idx}
                item={item}
                removeProductFromOrder={removeProductFromOrder}
              />
            ))}
          </tbody>
        }
      </TableOrderItem>
      {order?.orderItems && order.orderItems.length > 0 &&
        <div>
          <span>Total: </span>
          <span>{getTotalPrice()}$</span>
          <Button right={true} onClick={orderProduct} data-testid="order">Order</Button>
        </div>
      }
    </OrderList>
  </>);
}


const mapStateToProps = ({ orders }: IRootState) => ({
  actionStatus: orders.actionStatus
});

const mapDispatchToProps = {
  createOrder
}

type DispatchProps = typeof mapDispatchToProps;

type StateProps = ReturnType<typeof mapStateToProps>;

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Shop);
