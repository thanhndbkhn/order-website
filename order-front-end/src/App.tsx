import { BrowserRouter as Router, Route, Switch } from "react-router-dom";
import React from "react";
import Header from './Header';
import Shop from './components/modules/orders/shop/Shop';
import OrderList from './components/modules/orders/order/OrderList';
import styled from 'styled-components'

const Wrapper = styled.div`
  width: 70%;
  height: 100%;
  margin: 0 auto;
  padding: 10px;
  min-width: 768px;
  @media (max-width: 1024px) {
    margin-bottom: 20px;
    width: 100%;
    min-width: unset;
  }
`

const App: React.FC = () => {
  return (
    <Router>
      <Wrapper>
          <Header />
          <Switch>
            <Route path="/" exact component={Shop} />
            <Route path="/order" >
              <OrderList />
            </Route>
          </Switch>
      </Wrapper >
    </Router>
  );
}

export default App;
