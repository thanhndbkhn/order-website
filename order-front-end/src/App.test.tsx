import React from 'react';
import Adapter from "enzyme-adapter-react-16";
import App from './App';
import { shallow, configure } from 'enzyme';
import Shop from './components/modules/orders/shop/Shop';
import OrderList from './components/modules/orders/order/OrderList';
import OrderDetails from './components/modules/orders/order/OrderDetails';
import ReactDOM, { unmountComponentAtNode } from 'react-dom';
import { Provider } from 'react-redux';
import { applyMiddleware, createStore, Dispatch } from 'redux';
import rootReducer from './reducers/rootReducer.reducer';
import thunk from 'redux-thunk';
import { ACTION_TYPES, OrderAction } from './components/modules/orders/reducers/orders.reducers';
import { act } from "react-dom/test-utils";

let store: any = null;
let container: any = null;

beforeEach(() => {
  store = createStore(
    rootReducer,
    applyMiddleware(thunk)
  );
  configure({ adapter: new Adapter() });
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
  store = null;
});


it('renders App', () => {
  const div = document.createElement('div');
  ReactDOM.render(<Provider store={store}> <App /> </Provider>, div);
});

it('render List Order', () => {
  const responseMock =
  {
    data:
    {
      order: [
        {
          createdAt: "2021-03-04T15:46:30.257Z",
          orderId: "604100d6b4fbbb0023412d98",
          orderItems: [
            {
              product: {
                imageUrl: "https://demo.themeum.com/html/eshopper/images/home/product3.jpg",
                price: 300,
                productId: "3",
                productName: "Easy Polo 3"
              }, quantity: 1
            }],
          orderStatus: 3,
          totalPrice: 300
        }
      ]
    }
  }
  store.dispatch({
    type: ACTION_TYPES.GET_LIST_ORDER,
    response: responseMock,
  });
  const OrderListComponent = shallow(<Provider store={store}>
    <OrderList />
  </Provider>);
  expect(OrderListComponent).toMatchSnapshot();
});

it('render Order Detail', () => {
  const responseMock =
  {
    data:
    {
      createdAt: "2021-03-04T15:46:30.257Z",
      orderId: "604100d6b4fbbb0023412d98",
      orderItems: [
        {
          product: {
            imageUrl: "https://demo.themeum.com/html/eshopper/images/home/product3.jpg",
            price: 300,
            productId: "3",
            productName: "Easy Polo 3"
          }, quantity: 1
        }],
      orderStatus: 3,
      totalPrice: 300,
      userId: "1"
    }
  }

  store.dispatch({
    type: ACTION_TYPES.GET_DETAIL_ORDER,
    response: responseMock,
  });
  const orderDetailComponent = shallow(<Provider store={store}>
    <OrderDetails orderId={"604100d6b4fbbb0023412d98"} />
  </Provider>);
  expect(orderDetailComponent).toMatchSnapshot();
})

it('click order Order', () => {
  act(() => {
    ReactDOM.render(<Provider store={store}>
      <Shop />
    </Provider>, container);
  });

  // Click add product to Order
  const button = container.querySelector("[data-testid='1id']");
  expect(button?.innerHTML).toBe("Add to cart");
  act(() => {
    button?.dispatchEvent(new MouseEvent("click", { bubbles: true }));
  });

  // Get Button Order
  const buttonOrder = container.querySelector("[data-testid='order']");
  expect(buttonOrder?.innerHTML).toBe("Order");

  // Click button Order => Order Product1.
  act(() => {
    buttonOrder?.dispatchEvent(new MouseEvent("click", { bubbles: true }));
  });
})


it('renders Shop', () => {
  const responseMock =
  {
    message: 'Order Create Success',
  }
  store.dispatch({
    type: ACTION_TYPES.CREATE_ORDER,
    response: responseMock,
  });
  const shopComponent = shallow(<Provider store={store}>
    <Shop />
  </Provider>);
  expect(shopComponent).toMatchSnapshot();
});

it('click Name Order => Show Detail Order', () => {
  const responseMock =
  {
    data: [
        {
          createdAt: "2021-03-04T15:46:30.257Z",
          orderId: "604100d6b4fbbb0023412d98",
          orderItems: [
            {
              product: {
                imageUrl: "https://demo.themeum.com/html/eshopper/images/home/product3.jpg",
                price: 300,
                productId: "3",
                productName: "Easy Polo 3"
              }, quantity: 1
            }],
          orderStatus: 3,
          totalPrice: 300
        }
      ]
  }
  store.dispatch({
    type: ACTION_TYPES.GET_LIST_ORDER,
    response: responseMock,
  });
  act(() => {
    ReactDOM.render(<Provider store={store}>
      <OrderList />
    </Provider>, container);
  });
  // Name Product => show detail Order
  const listNameProduct = container.querySelector("[data-testid='604100d6b4fbbb0023412d98']");
  expect(listNameProduct.getElementsByTagName("a").item(0).textContent).toBe("Easy Polo 3");
})

import orderReducer from './components/modules/orders/reducers/orders.reducers'

it('Test Reducer Get order detail', () => {
  const responseMock = {
    data: {
        totalPrice: 300,
        userId: 1,
        orderItems: []
      }
  }
  const newStateOrder = orderReducer(undefined, {
    type: ACTION_TYPES.GET_DETAIL_ORDER,
    response: responseMock
  });
  expect(newStateOrder?.orderDetails).toEqual(responseMock.data);
})

it('Test Reducer Create Order', () => {
  const responseMock = {
    data: {
      totalPrice: 300,
      userId: 1,
      orderItems: []
    }
  }
  const newStateOrder = orderReducer(undefined, {
    type: ACTION_TYPES.CREATE_ORDER,
    response: responseMock
  });
  expect(newStateOrder.actionStatus.status).toEqual(OrderAction.CreateSuccess);
})

it('Test Reducer Cancel Order', () => {
  const order = {
    data: {
      totalPrice: 300,
      userId: 1,
      orderItems: []
    }
  }
  const newStateOrder = orderReducer(undefined, {
    type: ACTION_TYPES.CANCEL_ORDER,
    response: order
  });
  expect(newStateOrder.actionStatus.status).toEqual(OrderAction.CancelSuccess);
})
