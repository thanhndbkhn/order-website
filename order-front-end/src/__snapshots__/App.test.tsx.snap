// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render List Order 1`] = `ShallowWrapper {}`;

exports[`render Order Detail 1`] = `ShallowWrapper {}`;

exports[`renders Shop 1`] = `ShallowWrapper {}`;
